### PR TITLE
#6624: Exclude discarded tabs in broadcasts and other "all tabs" actions

### PR DESCRIPTION
**Pull request by fregante**
## What does this PR do?

- Part of https://github.com/pixiebrix/pixiebrix-extension/issues/6624

It's ready for review. ~~I'll want to test this before merging though~~ _Done_

## Checklist

- [x] Add tests
- [x] New files added to `src/tsconfig.strictNullChecks.json` (if possible)
- [x] Designate a primary reviewer: @grahamlangford

### CODE REVIEW

File: src/background/activeTab.ts
Comment (fregante) [inline src/background/activeTab.ts:1]: "activeTab" referred to some logic that was dropped a few months ago. I moved these functions to `extensionUtils`

File: src/contrib/google/sheets/core/useCurrentOrigin.ts
Comment (fregante) [inline src/contrib/google/sheets/core/useCurrentOrigin.ts:1]: The comments in this file seem copy-pasted and out of place (it mentions "in this PR", what PR?). I dropped some.

I ended up here looking for `tabs.query()` usage

File: src/background/executor.ts
Comment (fregante) [inline src/background/executor.ts:1]: Replacing `asyncForEach` with the more specific `forEachTab` here allowed for a slight refactoring, thanks to the included `Promise.allSettled`

File: src/background/executor.ts [hunk 1]
```diff
@@ -163,33 +162,23 @@ export async function requestRunInOtherTabs(
...
-    }
-  });
+  const { rejected, fulfilled } = groupPromisesByStatus(results);
 
-  if (rejected.size > 0) {
-    console.warn(`Broadcast rejected for ${rejected.size} tabs`, { rejected });
+  if (rejected.length > 0) {
+    console.warn(`Broadcast rejected for ${rejected.length} tabs`, {
+      rejected,
+    });
```
Comment (grahamlangford) [inline src/background/executor.ts:178]: What are the possible reasons for rejected tabs? Stale or closed tabs? Should we log this to Rollbar as well?

File: src/background/executor.ts [hunk 2]
```diff
@@ -199,30 +188,25 @@ export async function requestRunInAllFrames(
...
+  );
+
+  const { rejected, fulfilled } = groupPromisesByStatus(results);
 
-  if (rejected.size > 0) {
-    console.warn(`Broadcast rejected for ${rejected.size} frame`, { rejected });
+  if (rejected.length > 0) {
+    console.warn(`Broadcast rejected for ${rejected.length} frame`, {
+      rejected,
+    });
```
Comment (grahamlangford) [inline src/background/executor.ts:206]: Same question here: https://github.com/pixiebrix/pixiebrix-extension/pull/6626/files#r1354981167

**Comment by twschiller** [comment]
Adding @BLoe since he will be helping out with performance issues

File: src/background/executor.ts (hunk 1 above)
Comment (fregante) [reply src/background/executor.ts:178]: > What are the possible reasons for rejected tabs?

- [Non-scriptable tabs](https://github.com/fregante/webext-content-scripts#isscriptableurlurl). I can actually include a filter for this
- https://github.com/pixiebrix/pixiebrix-extension/issues/6548
- any other content script failure or timeout

> Stale or closed tabs?

Discarded tabs are excluded. There's a possible race condition if the user closes a tab between line 165 and line 166 😅 but I reckon it's rare.

> Should we log this to Rollbar as well?

I think this generally classifies as user/business error. We don't log errors that are due to unconnectable tabs.

https://github.com/pixiebrix/pixiebrix-extension/blob/3b9730a20839903c643928e9735a32fcc71da5ae/src/errors/errorHelpers.ts/#L69-L72

File: src/background/executor.ts (hunk 2 above)
Comment (fregante) [reply src/background/executor.ts:206]: Same as above, except that frames are more likely to come into existence and disappear. Still classifies as a non-reportable error.

**Review by grahamlangford — APPROVED** [review]

**Pull request merged**